pkg/mediatype/mediatype_of.go

```diff
@@ -0,0 +1,142 @@
+// sniffers to return a [MediaType] quickly before inspecting the content itself:
+// - Light Sniffing checks only the provided file extension or media type hints.
+// - Heavy Sniffing reads the bytes to perform more advanced sniffing.
+func mediaTypeOf(content SnifferContent, mediaTypes []string, fileExtensions []string, sniffers []Sniffer) *MediaType {
+
+	// Light sniffing with only media type hints
+	if len(mediaTypes) > 0 {
+		context := SnifferContext{
+			mediaTypes: mediaTypes,
+		}
+		for _, sniffer := range sniffers {
+			mediaType := sniffer(context)
+			if mediaType != nil {
+				return mediaType
+			}
+		}
+	}
+
+	// Light sniffing with both media type hints and file extensions
+	if len(fileExtensions) > 0 {
+		context := SnifferContext{
+			mediaTypes:     mediaTypes,
+			fileExtensions: fileExtensions,
+		}
+		for _, sniffer := range sniffers {
+			mediaType := sniffer(context)
+			if mediaType != nil {
+				return mediaType
+			}
+		}
+	}
+
+	// Heavy sniffing
+	if content != nil {
+		context := SnifferContext{
+			content:        content,
+			mediaTypes:     mediaTypes,
+			fileExtensions: fileExtensions,
+		}
+		for _, sniffer := range sniffers {
+			mediaType := sniffer(context)
+			if mediaType != nil {
+				return mediaType
+			}
+		}
+	}
+
+	// Falls back on the system-wide registered media types.
+	// Note: This is done after the heavy sniffing of the provided [sniffers], because
+	// otherwise it will detect JSON, XML or ZIP formats before we have a chance of sniffing
+	// their content (for example, for RWPM).
+	context := SnifferContext{
+		content:        content,
+		mediaTypes:     mediaTypes,
+		fileExtensions: fileExtensions,
+	}
+	if c := SniffSystem(context); c != nil {
+		return c
+	}
+
+	// If nothing else worked, we try to parse the first valid media type hint.
+	for _, mediaType := range mediaTypes {
+		if mediaType == "" {
+			continue // Blank mediatype
+		}
+		mt, err := NewMediaType(mediaType, "", "")
+		if err == nil {
+			return &mt
+		}
+	}
+
+	return nil
+}
+
+// Resolves a format from a list of mediatypes, list of extensions, and list of sniffers
+func MediaTypeOf(mediaTypes []string, extensions []string, sniffers []Sniffer) *MediaType {
+	return mediaTypeOf(nil, mediaTypes, extensions, sniffers)
+}
+
+func MediaTypeOfStringAndExtension(mediaType string, extension string) *MediaType {
+	return mediaTypeOf(nil, []string{mediaType}, []string{extension}, Sniffers)
+}
+
+// Resolves a format from a single mediaType string
+func MediaTypeOfString(mediaType string) *MediaType {
+	return mediaTypeOf(nil, []string{mediaType}, nil, Sniffers)
+}
+
+// Resolves a format from a single file extension
+func MediaTypeOfExtension(extension string) *MediaType {
+	return mediaTypeOf(nil, nil, []string{extension}, Sniffers)
+}
+
+// Resolves a format from a file
+func MediaTypeOfFile(file *os.File, mediaTypes []string, extensions []string, sniffers []Sniffer) *MediaType {
```

Comment:
In almost all cases we won't need to customize the `sniffers` when calling these functions. So let's simplify a bit and make it the default. We can add a `OfFileWithSniffers()`.

```suggestion
func OfFile(file *os.File, mediaTypes []string, extensions []string) *MediaType {
```